iavl_versioned_tree.go

```diff
@@ -0,0 +1,167 @@
+	versions       map[uint64]*orphaningTree // The previous, saved versions of the tree.
+	latestVersion  uint64                    // The latest saved version.
+	ndb            *nodeDB
+}
+
+// NewVersionedTree returns a new tree with the specified cache size and datastore.
+func NewVersionedTree(cacheSize int, db dbm.DB) *VersionedTree {
+	ndb := newNodeDB(cacheSize, db)
+	head := &IAVLTree{ndb: ndb}
+
+	return &VersionedTree{
+		orphaningTree: newOrphaningTree(head),
+		versions:      map[uint64]*orphaningTree{},
+		ndb:           ndb,
+	}
+}
+
+// LatestVersion returns the latest saved version of the tree.
+func (tree *VersionedTree) LatestVersion() uint64 {
+	return tree.latestVersion
+}
+
+// Tree returns the current working tree.
+func (tree *VersionedTree) Tree() *IAVLTree {
+	return tree.orphaningTree.IAVLTree
+}
+
+// String returns a string representation of the tree.
+func (tree *VersionedTree) String() string {
+	return tree.ndb.String()
+}
+
+// Load a versioned tree from disk. All tree versions are loaded automatically.
+func (tree *VersionedTree) Load() error {
+	roots, err := tree.ndb.getRoots()
+	if err != nil {
+		return err
+	}
+
+	// Load all roots from the database.
+	for _, root := range roots {
+		t := newOrphaningTree(&IAVLTree{ndb: tree.ndb})
+		t.Load(root)
+
+		version := t.root.version
+		tree.versions[version] = t
+
+		if version > tree.latestVersion {
+			tree.latestVersion = version
+		}
+	}
+	// Set the working tree to a copy of the latest.
+	tree.orphaningTree = tree.versions[tree.latestVersion].Clone()
+
+	return nil
+}
+
+// GetVersioned gets the value at the specified key and version.
+func (tree *VersionedTree) GetVersioned(key []byte, version uint64) (
+	index int, value []byte, exists bool,
+) {
+	if t, ok := tree.versions[version]; ok {
+		return t.Get(key)
+	}
+	return -1, nil, false
+}
+
+// SaveVersion saves a new tree version to disk, based on the current state of
+// the tree. Multiple calls to SaveVersion with the same version are not allowed.
+func (tree *VersionedTree) SaveVersion(version uint64) ([]byte, error) {
+	if _, ok := tree.versions[version]; ok {
+		return nil, errors.Errorf("version %d was already saved", version)
+	}
+	if tree.root == nil {
+		return nil, ErrNilRoot
+	}
+	if version == 0 {
+		return nil, errors.New("version must be greater than zero")
+	}
+	if version <= tree.latestVersion {
+		return nil, errors.New("version must be greater than latest")
+	}
+
+	tree.latestVersion = version
+	tree.versions[version] = tree.orphaningTree
+
+	tree.orphaningTree.SaveVersion(version)
+	tree.orphaningTree = tree.orphaningTree.Clone()
+
+	tree.ndb.SaveRoot(tree.root, version)
+	tree.ndb.Commit()
+
+	return tree.root.hash, nil
+}
+
+// DeleteVersion deletes a tree version from disk. The version can then no
+// longer be accessed.
+func (tree *VersionedTree) DeleteVersion(version uint64) error {
+	if version == 0 {
+		return errors.New("invalid version")
```

Comment:
"version must be greater than zero".

Below errors could include values too

Reply:
0215259